metrics/prometheus-metrics/src/main/scala/sttp/tapir/metrics/prometheus/PrometheusMetrics.scala

```diff
@@ -0,0 +1,136 @@
+package sttp.tapir.metrics.prometheus
+
+import io.prometheus.client.CollectorRegistry.defaultRegistry
+import io.prometheus.client.exporter.common.TextFormat
+import io.prometheus.client.{CollectorRegistry, Counter, Gauge, Histogram}
+import sttp.monad.MonadError
+import sttp.tapir.CodecFormat.TextPlain
+import sttp.tapir._
+import sttp.tapir.metrics.Metric
+import sttp.tapir.model.{ServerRequest, ServerResponse}
+import sttp.tapir.server.ServerEndpoint
+import sttp.tapir.server.interceptor.Interceptor
+import sttp.tapir.server.interceptor.metrics.MetricsInterceptor
+
+import java.io.StringWriter
+import scala.concurrent.duration.Deadline
+
+case class PrometheusMetrics[F[_]](registry: CollectorRegistry, metrics: Set[Metric[F, _]] = Set.empty[Metric[F, _]])(implicit
+    monad: MonadError[F]
+) {
+  import PrometheusMetrics._
+
+  private lazy val metricsEp = endpoint.get.in("metrics").out(plainBody[CollectorRegistry])
+
+  def withRequestsTotal: PrometheusMetrics[F] = copy(metrics = metrics + requestsTotal(registry))
+  def withRequestsActive: PrometheusMetrics[F] = copy(metrics = metrics + requestsActive(registry))
+  def withRequestsFailure: PrometheusMetrics[F] = copy(metrics = metrics + requestsFailures(registry))
+  def withResponsesTotal: PrometheusMetrics[F] = copy(metrics = metrics + responsesTotal(registry))
+  def withResponsesDuration: PrometheusMetrics[F] = copy(metrics = metrics + responsesDuration(registry))
+  def withCustom(m: Metric[F, _]): PrometheusMetrics[F] = copy(metrics = metrics + m)
+
+  def metricsServerEndpoint: ServerEndpoint[Unit, Unit, CollectorRegistry, Any, F] =
+    metricsEp.serverLogic { _ => monad.unit(Right(registry)) }
+
+  def metricsInterceptor[B](ignoreEndpoints: Seq[Endpoint[_, _, _, _]] = Seq.empty): Interceptor[F, B] =
+    new MetricsInterceptor[F, B](metrics.toList, ignoreEndpoints :+ metricsEp)
+}
+
+object PrometheusMetrics {
+
+  implicit val schemaForCollectorRegistry: Schema[CollectorRegistry] = Schema.string[CollectorRegistry]
+
+  implicit val collectorRegistryCodec: Codec[String, CollectorRegistry, CodecFormat.TextPlain] =
+    Codec.anyStringCodec(TextPlain())(_ => DecodeResult.Value(new CollectorRegistry()))(r => {
+      val output = new StringWriter()
+      TextFormat.write004(output, r.metricFamilySamples)
+      output.close()
+      output.toString
+    })
+
+  def withDefaultMetrics[F[_]](implicit monad: MonadError[F]): PrometheusMetrics[F] =
+    PrometheusMetrics[F](
+      defaultRegistry,
+      Set(
+        requestsTotal(defaultRegistry),
+        requestsActive(defaultRegistry),
+        requestsFailures(defaultRegistry),
+        responsesTotal(defaultRegistry),
+        responsesDuration(defaultRegistry)
+      )
+    )
+
+  def requestsTotal[F[_]](registry: CollectorRegistry)(implicit monad: MonadError[F]): Metric[F, Counter] =
+    Metric[F, Counter](
+      Counter
+        .build()
+        .namespace("tapir")
```

Comment:
These should be somehow configurable - e.g. for most apps this namespace doesn't make sense. It would be best if each metric would coudl be configured taking in the request/endpoint, and yielding a set of labels & label values to apply. With sane defaults, of course, so that adding default metrics is just a couple lines of code